### PR TITLE
xds: use order-deterministic locality map for preserving most recent locality update to eliminate test flakiness

### PR DESCRIPTION
**Pull request by voidzcy**
`LocalityStore` has clear-box tests such that inter-locality subchannel picking relies on the order of most recent locality update. Existing implementation loses the ordering when localities are stored in a `HashMap`, the order of `WeightedChildPicker`s (which are wrapped inter-localityo pickers) in `InterLocalityPicker` become indeterministic. This would cause test flakiness as the list of `WeightedChildPicker` passed into `FakePickerFactory#picker(...)` does not reflect the order of localities in most recent locality update.

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 1]
```diff
@@ -84,7 +83,7 @@
     private final StatsStore statsStore;
     private final OrcaPerRequestUtil orcaPerRequestUtil;
 
-    private Map<XdsLocality, LocalityLbInfo> localityMap = new HashMap<>();
+    private Map<XdsLocality, LocalityLbInfo> localityMap = new LinkedHashMap<>();
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/LocalityStore.java]: `ImmutableMap.of()`

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 2]
```diff
@@ -177,28 +176,22 @@ public void reset() {
         localityMap.get(locality).shutdown();
         statsStore.removeLocality(locality);
       }
-      localityMap = new HashMap<>();
+      localityMap = new LinkedHashMap<>();
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/LocalityStore.java]: ImmutableMap.of()

File: xds/src/main/java/io/grpc/xds/LocalityStore.java [hunk 3]
```diff
@@ -241,6 +233,7 @@ public void updateLocalityStore(Map<XdsLocality, LocalityInfo> localityInfoMap)
         }
         newState = aggregateState(newState, childHelper.currentChildState);
       }
+      localityMap = updatedLocalityMap;
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/LocalityStore.java]: `Collections.unmodifiableMap()`

File: xds/src/main/java/io/grpc/xds/LocalityStore.java (hunk 2 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Done.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java (hunk 3 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Done.

File: xds/src/main/java/io/grpc/xds/LocalityStore.java (hunk 1 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/LocalityStore.java]: Done.

**Comment by voidzcy** [comment]
> LGTM.
> Making code more "deterministic" and randomness come only from `Random` is always good.
> 
> Some comments for only if there is no mutation (I didn't thoroughly check).

Yes, it is immutable.

**Pull request merged**